feat: a gitfilesystem

nilleb opened the pull request:
The consumer of the VersionedCobertura has the possibility to inspect a coverage report from the past along with the corresponding sources.

Usage:
```python
VersionedCobertura(report, source=source, commit_id=commit_id)
```
Where `report` is the Cobertura report, `source` is the folder where the Git repository can be found and `commit` is the commit ID to which the report is associated.

aconrad commented on pycobertura/filesystem.py:
```diff
@@ -68,3 +69,52 @@ def open(self, filename):
+
+    def real_filename(self, filename):
+        prefix = "{}/".format(self.prefix) if self.prefix else ""
+        return "{p.commit_id}:{prefix}{filename}".format(
+            prefix=prefix, p=self, filename=filename
+        )
```
Can we instead do:

```suggestion
        return "{commit_id}:{prefix}{filename}".format(
            prefix=prefix, commit_id=self.commit_id, filename=filename
        )
```

... ? Would that work?

nilleb replied:
Yeah, of course. When I first wrote this I was playing with format strings, and unfortunately this is still in the code.

aconrad commented on pycobertura/filesystem.py:
```diff
@@ -68,3 +69,52 @@ def open(self, filename):
+
+
+class GitFileSystem(FileSystem):
+    def __init__(self, repo_folder, commit_id=None):
```
Should we make `commit_id` required? It seems that we're using it unconditionally.

```suggestion
    def __init__(self, repo_folder, commit_id):
```

nilleb replied:
👍

nilleb commented on pycobertura/filesystem.py:
```diff
@@ -68,3 +69,52 @@ def open(self, filename):
+class GitFileSystem(FileSystem):
+    def __init__(self, repo_folder, commit_id=None):
+        self.repository = repo_folder
+        self.commit_id = commit_id
```
Of course! _Bien vu_ ;-)

nilleb commented on pycobertura/filesystem.py:
```diff
@@ -68,3 +69,52 @@ def open(self, filename):
             raise self.FileNotFound(filename)
+
+
+def get_root_path(repository_folder):
```
Sorry for having just copied and pasted some code from elsewhere. I agree with what you say, it totally makes sense in this organization of code.

nilleb commented on pycobertura/utils.py:
```diff
@@ -208,3 +210,15 @@ def hunkify_lines(lines, context=3):
+    except OSError:
+        logging.error("Command being executed: {}".format(command))
+        raise
+
```
I have noticed that in both my pull requests I left trailing whitespaces.

Would you mind if I use black to format sources? I gave it a try on your code (that is very nicely formatted), and if you were OK to switch to a line length of 88 characters, it would, I think, fit your needs.

aconrad replied:
I've never used black before. Also the homepage doesn't make it clear to what it covers. It seems like it would work like "prettier", no? Does it work only for Python?

aconrad replied:
If we were to use Black, I'd rather do this in a separate PR if you don't mind.

nilleb commented:
I have also squeezed the `get_output` trying to rationalize the use of `subprocess` in `filesystem.py`.

Concerning the unit tests: of course, sorry. I am going to provide a distinct commit.
I have just seen that some unit tests are failing on python 27. Would you consider dropping unsupported versions of python from tox.ini? :-)

Concerning the trailing whitespaces: I have noticed that in both my pull requests I left some of them.
Would you mind if I use black to format sources? I gave it a try on your code (that is very nicely formatted), and if you were OK to switch to a line length of 88 characters, it would, I think, fit your needs. I will try to provide this in a distinct commit, so that if you are not OK I'll squeeze it out.

And, last, concerning the design patterns, yeah, I totally agree. The Cobertura __init__ is a factory method of a filesystem. The choice you have made was the best one, because simplicity is at stake. For the future, a constructor with a FileSystem argument instead of the existing ones, as you have suggested, could be the best one to improve code readability and single responsibility principle.

If you are OK to major bump, I'll propose a rework following your suggestions.

aconrad commented:
Happy to drop unsupported versions of python. We might as well do that if we're going to do a major version bump.

In addition, we should check if the readme needs to be updated accordingly with the changes you are about to make. And add the changes in the change log file, under "unreleased".

aconrad commented on tests/test_filesystem.py:
```diff
@@ -96,3 +96,52 @@ def test_filesystem_zip__with_source_prefix():
+        with fs.open(filename) as f:
+            assert hasattr(f, 'read')
+
+        assert git_filename.startswith(branch)
```
Is it possible to assert for the whole filename just to be sure it contains everything necessary? For example:

```suggestion
        expected_git_filename = "master:tests/dummy/test-file"
        assert git_filename == expected_git_filename 
```

aconrad commented on tox.ini:
```diff
@@ -1,6 +1,6 @@
 [tox]
 skipsdist = True
-envlist = py26, py27, py34, pep8
+envlist = py34, py37, py38, pep8
```
Do we want to support py35 and/or py36? It has been a few years since I've worked on Python professionally so please excuse me if I'm a little rusty on my understanding of the Python ecosystem.

aconrad replied:
Also, I'm seeing here that Python 3.4 has reached end-of-life. Should we remove it from tox as well? https://www.python.org/downloads/release/python-340/

aconrad commented on pycobertura/filesystem.py:
```diff
@@ -68,3 +71,52 @@ def open(self, filename):
+        """
+        filename = self.real_filename(filename)
+
+        command = "git --no-pager show {}".format(filename)
+
+        try:
+            output = subprocess.check_output(shlex.split(command), cwd=self.repository)
```
Yay for using `subprocess`!

aconrad replied:
Non-blocking: Is it possible that `shlex.split()` fails here? Should we move it outside of the try/except block so we reduce the scope of what is being executed within try/except?

```suggestion
        command = "git --no-pager show {}".format(filename)
        command = shlex.split(command)

        try:
            output = subprocess.check_output(command, cwd=self.repository)
```

aconrad commented on pycobertura/filesystem.py:
```diff
@@ -68,3 +71,52 @@ def open(self, filename):
+
+        try:
+            output = subprocess.check_output(shlex.split(command), cwd=self.repository)
+        except (OSError, subprocess.CalledProcessError):
```
Do we still need `OSError` here now that we use `subprocess`? In what circumstances would we expect `OSError` to get raised?

nilleb replied:
Well, according to https://docs.python.org/2/library/subprocess.html#exceptions
we should catch OSError (on systems where git is not available) and CalledProcessError (if the executable returns a non-zero error code).
I think this is a good idea to keep both.

aconrad replied:
Oh great, thanks for looking into it!

aconrad commented on pycobertura/filesystem.py:
```diff
@@ -68,3 +71,52 @@ def open(self, filename):
+        )
+        return not bool(return_code)
+
+    def _get_root_path(self, repository_folder):
+        command = "git rev-parse --show-toplevel"
+        output = subprocess.check_output(shlex.split(command), cwd=repository_folder)
+        return output.decode("utf-8").rstrip()
```
Yay, thanks for making this method scoped to the class!

aconrad commented on tests/test_filesystem.py:
```diff
@@ -96,3 +96,52 @@ def test_filesystem_zip__with_source_prefix():
+        subprocess_mock.check_output = MagicMock(return_value=b"<file-content>")
+
+        fs = GitFileSystem(folder, branch)
+        git_filename = fs.real_filename(filename)
```
Can we move this line closer to where it is being used a few lines below?

aconrad commented on tests/test_filesystem.py:
```diff
@@ -96,3 +96,52 @@ def test_filesystem_zip__with_source_prefix():
+def test_filesystem_git_integration():
+    from pycobertura.filesystem import GitFileSystem
+
+    fs = GitFileSystem(".", "d1fe88da6b18340762b24bb1f89067a3439c4041")
```
Can we add a little explicitness as to what this commit is? Without a note, it leaves the reader wondering where that commit comes from.

```suggestion
    first_pycobertura_commit = "d1fe88da6b18340762b24bb1f89067a3439c4041"
    fs = GitFileSystem(".", first_pycobertura_commit)
```

Alternatively, we could also make it a CONSTANT at the top of this module and reuse the constant throughout this file:

```python
FIRST_PYCOBERTURA_COMMIT_SHA = "d1fe88da6b18340762b24bb1f89067a3439c4041"
```

... and then

```python
fs = GitFileSystem(".", FIRST_PYCOBERTURA_COMMIT_SHA)
```

What do you think?

aconrad commented on tests/test_filesystem.py:
```diff
@@ -96,3 +96,52 @@ def test_filesystem_zip__with_source_prefix():
+def test_filesystem_git_integration__not_found():
+    from pycobertura.filesystem import GitFileSystem
+
+    fs = GitFileSystem(".", "d1fe88da6b18340762b24bb1f89067a3439c4041")
```
Same here.

```suggestion
    first_pycobertura_commit = "d1fe88da6b18340762b24bb1f89067a3439c4041"
    fs = GitFileSystem(".", first_pycobertura_commit)
```

You could also make it a CONSTANT at the top of this module:

```python
FIRST_PYCOBERTURA_COMMIT_SHA = "d1fe88da6b18340762b24bb1f89067a3439c4041"
```

... and then

```python
fs = GitFileSystem(".", FIRST_PYCOBERTURA_COMMIT_SHA)
```

aconrad commented on tests/test_filesystem.py:
```diff
@@ -96,3 +96,52 @@ def test_filesystem_zip__with_source_prefix():
+
+    fs = GitFileSystem(".", "d1fe88da6b18340762b24bb1f89067a3439c4041")
+
+    source_files = [
```
Non-blocking: Can we add a small comment here to understand why we picked these files to test on?

```suggestion
    # Files included in pycobertura's first commit.
    source_files = [
```

aconrad commented on pycobertura/filesystem.py:
```diff
@@ -68,3 +71,52 @@ def open(self, filename):
+    def has_file(self, filename):
+        command = "git --no-pager show {}".format(self.real_filename(filename))
+        return_code = subprocess.call(
+            command,
```
Should we be using `shlex` here, too?

aconrad commented on pycobertura/filesystem.py:
```diff
@@ -68,3 +71,52 @@ def open(self, filename):
+        self.repository = repo_folder
+        self.ref = ref
+        self.repository_root = self._get_root_path(repo_folder)
+        self.prefix = self.repository.replace(self.repository_root, "").lstrip("/")
```
Can you elaborate on how `prefix` is used in the context of `GitFileSystem`? It's not immediately obvious by reading the code.

nilleb commented:
I have addressed your latest review comments, but I have not yet changed the unreleased section of the changelog.
This because I am trying to add a last commit that will change the signature of the Cobertura constructor, as you suggested in https://github.com/aconrad/pycobertura/pull/94#pullrequestreview-396054566.
But if you prefer to keep the breaking changes for later, I'll just update the changelog.

aconrad commented on pycobertura/cobertura.py:
```diff
@@ -41,37 +36,25 @@ class Cobertura(object):
-            self.filesystem = DirectoryFileSystem(
-                source, source_prefix=source_prefix
-            )
+    @property
+    def filesystem(self):
+        if not self._filesystem:
+            self._filesystem = filesystem_factory(self._report)
+        return self._filesystem
```
What is the reason for adding this getter? Should we change the references in this file to access `self._filesystem` instead?

nilleb replied:
Because when the Cobertura instance has been built with `Cobertura(xml)` there is no filesystem at all, and the two methods `file_source` and `source_lines` could throw. This property could be simply replaced with a 
`self.filesystem = filesystem if filesystem else filesystem_factory(report)` at line 51.
Or just dropped if you want your consumers to get exceptions when no filesystem has been passed at all.

aconrad replied:
Gotcha.

Let's remove it as I would rather not expose this attribute as part of the public API. In hindsight, I should've made it internal/private in the first place. Also, now that the filesystem is being passed by the caller, they already have access to it should if they need the instance.

I would rather let it die or throw an exception that says that the Cobertura instance wasn't configured with a filesystem to be friendlier (but you don't have to do it, I could add that in a subsequent PR).

aconrad commented on pycobertura/filesystem.py:
```diff
@@ -68,3 +71,106 @@ def open(self, filename):
+
+    if ref:
+        return GitFileSystem(source, ref)
+    else:
+        return DirectoryFileSystem(source, source_prefix=source_prefix)
```
We can avoid the else-statement here since we use a guard-clause above.

```suggestion

    return DirectoryFileSystem(source, source_prefix=source_prefix)
```

aconrad commented on pycobertura/filesystem.py:
```diff
@@ -68,3 +71,106 @@ def open(self, filename):
+            source, source_prefix=source_prefix
+        )
+
+    if os.path.isfile(source):
+        source = os.path.dirname(source)
```
The keyword argument `source` is expected to be a directory according to the documentation, so I'm not sure when `source` could be a file path. In what scenario would this if-statement be true?

aconrad commented on tests/test_filesystem.py:
```diff
@@ -96,3 +99,55 @@ def test_filesystem_zip__with_source_prefix():
+    with patch.object(fsm, "subprocess") as subprocess_mock:
+        subprocess_mock.check_output = MagicMock(return_value=b"<file-content>")
+
+        fs = GitFileSystem(folder, branch)
```
Non-blocking: could we reuse the import `fsm` and avoid the import `from pycobertura.filesystem import GitFileSystem` above?

```suggestion
        fs = fsm.GitFileSystem(folder, branch)
```

aconrad commented on CHANGES.md:
```diff
@@ -2,6 +2,9 @@
 
 ## Unreleased
 
+* Add a GitFilesystem, allowing the consumer to retrieve the sources from a git commit.
```
Let's get you some credit for the work you have done here! 🎉

```suggestion
* Add a `GitFilesystem` to allow pycobertura to access source files at different revisions from a git repository. Thanks @nilleb.
```

nilleb commented:
Given the times we live in, I hope that you are doing well. I just wanted to check how did you go. Have a good day!
Ivo

aconrad commented:
Hey @nilleb, thanks for checking in! I'm all good. I have a lot going on right now (new job search due to covid-19) which is why I've been a little slower to respond.

TLDR: I think your PR looks great! I'm ready to merge if we can bring up the coverage if `pycobertura/filesystem.py` back to 100% coverage. Can we do that?

_Below is a brain dump of some early investigation as to why the coverage is so off between py37 and py38._

---------------------

I ran your PR locally but I'm seeing the coverage report being really off. And I'm not sure why but it seems to only affect Python 3.8, prior versions of it work fine...

Here is the coverage report with py37 (against this PR):

```
===================================== warnings summary =====================================
.tox/py37/lib/python3.7/site-packages/_pytest/junitxml.py:417
  /mnt/c/Users/Alex/Documents/work/pycobertura/.tox/py37/lib/python3.7/site-packages/_pytest/junitxml.py:417: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0.
  Add 'junit_family=xunit1' to your pytest.ini file to keep the current format in future versions of pytest and silence this warning.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
-------- generated xml file: /mnt/c/Users/Alex/Documents/work/pycobertura/junit.xml --------

----------- coverage: platform linux, python 3.7.5-final-0 -----------
Name                                Stmts   Miss  Cover   Missing
-----------------------------------------------------------------
pycobertura/__init__.py                 2      0   100%
pycobertura/cli.py                     64      0   100%
pycobertura/cobertura.py              190      0   100%
pycobertura/filesystem.py              88      7    92%   94-103, 112-113, 171
pycobertura/reporters.py              164      0   100%
pycobertura/templates/__init__.py       0      0   100%
pycobertura/templates/filters.py        8      0   100%
pycobertura/utils.py                  111      1    99%   31
-----------------------------------------------------------------
TOTAL                                 627      8    99%
Coverage XML written to file coverage.xml

============================== 101 passed, 1 warning in 4.10s ==============================
_________________________________________ summary __________________________________________
  py37: commands succeeded
  congratulations :)
```

(NOTE: do you think you can get `pycobertura/filesystem.py` back to 100% coverage?)

... and here is the one against py38:

```
===================================== warnings summary =====================================
.tox/py38/lib/python3.8/site-packages/_pytest/junitxml.py:417
  /mnt/c/Users/Alex/Documents/work/pycobertura/.tox/py38/lib/python3.8/site-packages/_pytest/junitxml.py:417: PytestDeprecationWarning: The 'junit_family' default value will change to 'xunit2' in pytest 6.0.
  Add 'junit_family=xunit1' to your pytest.ini file to keep the current format in future versions of pytest and silence this warning.
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
-------- generated xml file: /mnt/c/Users/Alex/Documents/work/pycobertura/junit.xml --------

----------- coverage: platform linux, python 3.8.0-final-0 -----------
Name                                Stmts   Miss  Cover   Missing
-----------------------------------------------------------------
pycobertura/__init__.py                 2      0   100%
pycobertura/cli.py                    109     81    26%   268-3726
pycobertura/cobertura.py              203     35    83%   315, 416-911
pycobertura/filesystem.py             103     26    75%   94-96, 339-886
pycobertura/reporters.py              207     99    52%   276, 289, 299, 315, 318, 320-1314
pycobertura/templates/__init__.py       0      0   100%
pycobertura/templates/filters.py       13      8    38%   261-536
pycobertura/utils.py                  111      1    99%   31
-----------------------------------------------------------------
TOTAL                                 748    250    67%
Coverage XML written to file coverage.xml

============================== 101 passed, 1 warning in 3.83s ==============================
_________________________________________ summary __________________________________________
  py38: commands succeeded
  congratulations :)
```

And this file definitely doesn't have a line number "3726" in `cli.py`:

```
pycobertura/cli.py                    109     81    26%   268-3726
```

As you can see above there is a deprecation warning about `xunit1` vs. `xunit2`. I ran a `pip freeze` between both versions of py37 and py38 but the versions of `coverage` and `pytest` seem to be the same in both pyenvs.

![image](https://user-images.githubusercontent.com/247661/80516623-963d6000-8938-11ea-8b74-ea04f039d672.png)

So that _seems_ unrelated. I wonder if Python 3.7 and 3.8 measure line hits/misses in a different way... ? Maybe we should try to upgrade `coverage` and `pytest-cov` and have pycobertura work on more recent versions... `¯\_(ツ)_/¯`

A while ago, we pinned versions of `coverage` and `pytest-cov` in `test-requirements.txt` because later versions were causing pycobertura to fail due to problematic data in the generated coverage file:

```
cat test-requirements.txt | grep cov
coverage>=4.0,<4.4
pytest-cov<2.6.0
```

I found the ticket that required me to pin to older versions here: https://github.com/nedbat/coveragepy/issues/578 (the original report was on bitbucket here: https://bitbucket.org/ned/coveragepy/issues/578/incomplete-file-path-in-xml-report)

nilleb commented:
You should be good to go.

I had no problem removing the version spec from coverage and pytest, at least not with py37 and py38, and this has brought the _missing lines_ to a valid interval.

I have added an option to avoid the pytest warning to the setup.cfg.

But I have reverted (dropped) a call to shlex, because it prevented the subprocess.call from functioning.

```
----------------------------------- generated xml file: /Users/ivo/dev/pycobertura/junit.xml ------------------------------------

---------- coverage: platform darwin, python 3.8.2-final-0 -----------
Name                                Stmts   Miss  Cover   Missing
-----------------------------------------------------------------
pycobertura/__init__.py                 2      0   100%
pycobertura/cli.py                     66      0   100%
pycobertura/cobertura.py              199      0   100%
pycobertura/filesystem.py              90      0   100%
pycobertura/reporters.py              164      0   100%
pycobertura/templates/__init__.py       0      0   100%
pycobertura/templates/filters.py        8      0   100%
pycobertura/utils.py                  111      1    99%   31
-----------------------------------------------------------------
TOTAL                                 640      1    99%
Coverage XML written to file coverage.xml
```

nilleb commented:
And, good luck for job hunt.

aconrad commented:
Thanks, @nilleb, for this wonderful PR. I don't think we're quite ready to unpin the test requirements after all but I wanted to get this off your plate so I can keep investigating on my side.
